polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp

```diff
@@ -1056,8 +1056,90 @@ struct EraseSPIRVBuiltinPattern
   }
 };
 
-struct ConvertPolygeistToLLVMPass
+class DuplicateFuncCleaner {
+public:
+  struct Entry {
+    Entry(StringRef funcName, FunctionType funcType)
+        : funcName(funcName), signature(funcType) {}
+
+    StringRef funcName;
+    FunctionType signature;
+    bool anyHadConflictingSignature = false;
+    bool found = false;
+  };
+
+  explicit DuplicateFuncCleaner(std::initializer_list<Entry> init)
+      : entries(init) {
+    llvm::sort(entries, [](const Entry &lhs, const Entry &rhs) {
+      return lhs.funcName < rhs.funcName;
+    });
+  }
+
+  LogicalResult run(ModuleOp module) {
+    std::vector<LLVM::LLVMFuncOp> toRemove;
+    const auto getEntry = [this](LLVM::LLVMFuncOp func) {
+      auto iter = llvm::lower_bound(entries, func.getName(),
+                                    [](const Entry &entry, StringRef name) {
+                                      return name < entry.funcName;
+                                    });
+      if (iter == entries.end())
+        return iter;
+      return iter->funcName == func.getName() ? iter : entries.end();
+    };
+    for (auto func : module.getOps<LLVM::LLVMFuncOp>()) {
+      auto iter = getEntry(func);
```

Comment:
Nit: Replace `auto` with type.

Reply:
As it's `std::vector<LLVM::LLVMFuncOp>::iterator`, are we OK with leaving as is?

Reply:
Isn't this one an `Entry` or rather an iterator to `Entry`? 

If it's in iterator, can probably keep `auto` though.

Reply:
Changed code. No longer applies.